Separate code manager for the interpreter

janvorli opened the pull request:
This change adds a separate code manager for the interpreted code. Here is a high level overview of the changes:
* A base class `EECodeGenManager` was extracted from the `EEJitManager`
* A new class `InterpreterJitManager` derived from the `EECodeGenManager` was added and implemented.
* Similarly, a base class `CEECodeGenInfo` was extracted from the `CEEJitInfo`
* A new class `CInterpreterJitInfo` derived from the `CEECodeGenInfo` was added and implemented
* The `CodeHeader` and `RealCodeHeader` became base classes and `JitCodeHeader`, `RealJitCodeHeader`, `InterpreterCodeHeader` and `RealInterpreterCodeHeader` were added. The `CodeHeader` derivates don't add any extra data, the `JitCodeHeader` just provides the methods that are needed to access Jit specific members of the `RealJitCodeHeader`.
* The `UnsafeJitFunction` was refactored so that the interpreter (if enabled) is invoked first and then the AltJit / Jit ones are called.
* A new `DOTNET_Interpreter`, `DOTNET_InterpreterPath` and `DOTNET_InterpreterName` env vars are added to match the existing AltJit ones.

janvorli committed:
Separate code manager for the interpreter

This change adds a separate code manager for the interpreted code. Here
is an overview of the changes:
* A base class EECodeGenManager is extracted from the EEJitManager
* A new class InterpreterJitManager derived from the EECodeGenManager is
  added and implemented.
* Similarly, a base class CEECodeGenInfo is extracted from the CEEJitInfo
* A new class CInterpreterJitInfo derived from the CEECodeGenInfo is added
  and implemented
* The CodeHeader and RealCodeHeader became base classes and
  JitCodeHeader, RealJitCodeHeader, InterpreterCodeHeader and
  RealInterpreterCodeHeader were added. The CodeHeader derivates don't
  add any extra data, the JitCodeHeader just provides the methods that
  are needed to access Jit specific members of the RealJitCodeHeader.
* The UnsafeJitFunction is refactored so that the interpreter (if
  enabled) is invoked first and then the AltJit / Jit ones are called.
* A new DOTNET_Interpreter, DOTNET_InterpreterPath and
  DOTNET_InterpreterName env vars are added to match the existing AltJit
  ones.

janvorli commented:
cc: @BrzVlad

max-charlamb commented:
This conflicts with #111759 because cDAC starts to use the `RealCodeHeader` offsets for `nUnwindInfos` and `unwindInfos` that this PR moves to the `JitRealCodeHeader`. The cDAC does not currently support the interpreter so I think it only cares about the `JitRealCodeHeader` right now and we should change the datadescriptors to use that instead of the new base class. This would be a breaking cDAC change and require the DataType name to change on the managed side too, but there isn't any impact.

https://github.com/dotnet/runtime/pull/111759/files#diff-d10a483484c15e47b5c1d129ff70b1bb88ad775e8f113475f8789aad1a4ffad5

@davidwrighton

jkotas commented:
> This conflicts with https://github.com/dotnet/runtime/pull/111759 because cDAC starts to use the RealCodeHeader offsets for nUnwindInfos and unwindInfos that this PR moves to the JitRealCodeHeader

My preference would be to solve this by avoiding coupling between the JIT code manager and interpreter code manager. Leave the existing code header for JITed code alone, and define a dedicated code header for interpreted code. There is going to be some duplication for now and that's fine. The advantage is that it allows each of them to evolve independently.

jkotas commented:
> I am not sure I understand why cdac depends on the name of the code header

cdac infrastructure expects the data contract type representation to match the implementation. If you split a type into two types without actually changing the layout, it changes the data contract. You are right that the contract type representation does not have to match the implementation. There is probably a way how to convince the cdac infrastructure to deal with the divergence, but it is not how it is usually done. I am not sure whether we have a prior example.

> do you mean instead of the shared EECodeGenManager base class to have the InterpreterJitManager a copy of almost all of the code from the EEJitManager?

I am concerned about the code header. I think we want the interpreter bytecode header to be decoupled from JITed code header, so we are free to change one in any way we want without impacting the other.

I do not have a strong opinion whether it should be implemented as inheritance hierarchy, or as two types calling into static helpers to facilitate code sharing.

janvorli commented:
Ah, so the concern is just the code header and the code that works with it, that makes sense. I also think we may eventually diverge the interpreter and jit code header. I was thinking you were talking about all of the code in the code manager.

janvorli committed:
Make the code headers independent

Make the jit and interpreter code headers independent

janvorli committed:
Added missing #ifdef

am11 commented on src/coreclr/debug/daccess/enummem.cpp:
```diff
@@ -318,8 +318,12 @@ HRESULT ClrDataAccess::EnumMemDumpJitManagerInfo(IN CLRDataEnumMemoryFlags flags
+        EECodeGenManager* managerPtr = ExecutionManager::GetEEJitManager();
         managerPtr->EnumMemoryRegions(flags);
+#ifdef FEATURE_INTERPRETER
+        managerPtr = ExecutionManager::GetInterpreterJitManager();
```
Should it be using `#ifdef` `#else` blocks (to avoid overwriting the values)?

filipnavara replied:
It should enumerate both memory regions and it just reuses local variable. This seems fine.

am11 replied:
Ah, ok if enumerating both is the intention, it's fine.

janvorli replied:
No, it needs to call the enumeration for both managers

am11 commented on src/coreclr/vm/codeman.h:
```diff
@@ -2447,6 +2668,159 @@ class ReadyToRunJitManager final: public IJitManager
+    {
+        // Not used for the interpreter
+        _ASSERTE(FALSE);
+        return NULL;
```
```suggestion
        return (TADDR)NULL;
```

janvorli commented:
I've tried to run the dotnet/performance microbenchmarks for EH. All of them are about 5% worse with the last change that separates the code headers completely. With the initial change, the difference was < 2%. I need to think about some other way to solve the code header separation.

janvorli committed:
Fix MUSL build break

janvorli commented:
I have reworked it and got rid of all of the regression. I will update this PR on Monday

janvorli committed:
Rework of the CodeHeader separation

The previous way was introducing 5% regression in all exception handling
microbenchmarks, this gets us back to the same perf as without any code
manager changes.

janvorli committed:
Fix Unix build failure

The clang doesn't do late parsing of the template code, so the helpers
cannot be in the .inl file as they reference some types that are not
defined yet at that point.

janvorli commented:
@jkotas can you please take another look? The code headers are now completely separated.

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -3893,7 +4130,7 @@ BOOL EEJitManager::JitCodeToMethodInfo(
+    TADDR start = dac_cast<PTR_EECodeGenManager>(pRangeSection->_pjit)->FindMethodCode(pRangeSection, currentPC);
     if (start == (TADDR)0)
         return FALSE;
 
```
What happens if we end up with `InterpreterCodeHeader` in this and other similar places?

janvorli replied:
Thank you for spotting this. I have overlooked the usage of the `CodeHeader` explicitly in this method. It needs to be abstracted.

janvorli replied:
Hmm, I can see I have missed couple of other places too. Let me rescan the usage of the `CodeHeader` to make sure it doesn't appear where it should not.

janvorli replied:
Fixed

janvorli commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -4644,6 +4885,14 @@ BOOL ExecutionManager::IsManagedCodeWorker(PCODE currentPC, RangeSectionLockStat
 #endif
+#ifdef FEATURE_INTERPRETER
+    else
+    if (pRS->_flags & RangeSection::RANGE_SECTION_INTERPRETER)
```
Right, thank you for spotting this. It should be checked before the `RANGE_SECTION_CODEHEAP`. Actually, it looks like the JIT code path could use the `JitCodeToMethodInfo` too and all the three cases could be unified, calling the `JitCodeToMethodInfo` without casting to the specific jit manager classes. The `JitCodeToMethodInfo` is virtual.

janvorli commented on src/coreclr/debug/daccess/fntableaccess.h:
```diff
@@ -11,7 +11,7 @@
 
 
-struct FakeEEJitManager
+struct FakeEECodeGenManager
```
This was actually a leftover from the first version of the change. It is not needed.

janvorli committed:
PR feedback and some cleanup

janvorli commented on src/coreclr/inc/corjit.h:
```diff
@@ -52,6 +52,7 @@ enum CorJitAllocMemFlag
     CORJIT_ALLOCMEM_FLG_32BYTE_ALIGN   = 0x00000004, // The code will be 32-byte aligned
     CORJIT_ALLOCMEM_FLG_RODATA_32BYTE_ALIGN = 0x00000008, // The read-only data will be 32-byte aligned
     CORJIT_ALLOCMEM_FLG_RODATA_64BYTE_ALIGN = 0x00000010, // The read-only data will be 64-byte aligned
+    CORJIT_ALLOCMEM_FLG_INTERPRETED = 0x00000020,         // The code is interpreted byte code
```
Removed, was a remainder of the previous state.

janvorli committed:
Fix a bug

janvorli committed:
Fix build break

am11 commented on src/coreclr/vm/dynamicmethod.cpp:
```diff
@@ -437,19 +438,29 @@ HeapList* HostCodeHeap::InitializeHeapList(CodeHeapRequestInfo *pInfo)
+    else
+#endif // FEATURE_INTERPRETER
+    {
+#if defined(TARGET_AMD64) || defined(TARGET_ARM64) || defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
```
nit: this set is `TARGET_64BIT`.

janvorli replied:
Done

BrzVlad commented on src/tests/JIT/interpreter/InterpreterTester.cs:
```diff
@@ -22,9 +22,9 @@ public static void RunTests()
-        startInfo.EnvironmentVariables["DOTNET_AltJit"] = "RunInterpreterTests";
+        startInfo.EnvironmentVariables["DOTNET_InterpreterName"] = libInterp;
+        startInfo.EnvironmentVariables["DOTNET_InterpreterPath"] = Path.Combine(coreRoot, libInterp);
+        startInfo.EnvironmentVariables["DOTNET_Interpreter"] = "RunInterpreterTests";
```
When using the AltJit option I noticed I had problems unless I passed all these flags. Shouldn't as of this change be enough to just pass the `DOTNET_Interpreter`, while the rest could be omitted ?

janvorli replied:
Right, it should be sufficient. Let me get rid of the other env variables setting in the test.

janvorli replied:
Done

janvorli commented:
@jkotas I have realized I have not realized I also need to add an InterpreterCodeManager (ICodeManager derived one), so I am working on adding a one to this PR.

jkotas commented:
BTW: I do not think that the split between IJitManager and ICodeManager is very sensible. If we fine a good reason to, I think it would be fine to merge the two into a single interface.

janvorli committed:
Add dummy ICodeManager derived InterpreterCodeManager

janvorli committed:
Remove unused ICodeManager methods

davidwrighton commented on src/coreclr/vm/jitinterface.h:
```diff
@@ -565,18 +565,218 @@ class CEEInfo : public ICorJitInfo
 // CEEJitInfo is the concrete implementation of callbacks that the EE must provide for the JIT to do its
 // work.   See code:ICorJitInfo#JitToEEInterface for more on this interface.
-class CEEJitInfo : public CEEInfo
+class CEEJitInfo : public CEECodeGenInfo
```
Nit, I think you can use `class CEEJitInfo final : public CEECodeGenInfo` instead of scattering final throughout the definition of `CEEJitInfo`. Also the same for `CInterpreterJitInfo`

janvorli replied:
Right, a bulk of it was just a move / copy of the function headers from the original `CEEJitInfo` that had it per function. I will change it to per class except for the few ones that are final in the `CEECodeGenInfo` shared base class.

janvorli replied:
Done

janvorli committed:
PR feedback

janvorli commented:
@jkotas, @davidwrighton do you have any more comments on this PR?

janvorli commented:
@BrzVlad, @kg, can you please take a look too?

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -11023,7 +11039,7 @@ void CEEJitInfo::reportRichMappings(
     JIT_TO_EE_TRANSITION();
 
-    if (m_jitManager->IsStoringRichDebugInfo())
+    if (((EEJitManager*)m_jitManager)->IsStoringRichDebugInfo())
```
`IsStoringRichDebugInfo` lives on `EECodeGenManager` . Is this cast correct?

janvorli replied:
You are right, this is a remainder of an intermediate change that I have later reverted.

janvorli commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2821,13 +2906,30 @@ void EEJitManager::allocCode(MethodDesc* pMD, size_t blockSize, size_t reserveFo
+#endif
+        _ASSERTE(reserveForJumpStubs == 0);
+        requestInfo.SetInterpreted();
+        requestInfo.SetDynamicDomain();
```
Ah, thank you for spotting this. I have added it temporarily to test that allocation from dynamic domain works correctly too without having interpreted code that would need it and forgotten about it.

janvorli commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -416,7 +422,7 @@ void UnwindInfoTable::AddToUnwindInfoTable(UnwindInfoTable** unwindInfoPtr, PT_R
             // This cast is justified because only EEJitManager's have the code type above.
             EEJitManager* pJitMgr = (EEJitManager*)(pRS->_pjit);
-            CodeHeader * pHeader = pJitMgr->GetCodeHeaderFromStartAddress(entryPoint);
+            CodeHeader * pHeader = dac_cast<PTR_CodeHeader>(pJitMgr->GetCodeHeaderFromStartAddress(entryPoint));
```
Right

janvorli commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -3247,91 +3326,57 @@ void * EEJitManager::allocCodeFragmentBlock(size_t blockSize, unsigned alignment
-
-GCInfoToken EEJitManager::GetGCInfoToken(const METHODTOKEN& MethodToken)
+template<typename TCodeHeader>
+BYTE* EECodeGenManager::allocGCInfo(TCodeHeader* pCodeHeader, DWORD blockSize, size_t * pAllocationSize)
```
I will unify those allocations and move the actual setting to the callers for both.

janvorli committed:
PR feedback

* Refactored GC and EH info allocation
* Added Interpreter-TODO at some places
* Reverted some changes that were not needed anymore

janvorli commented:
@jkotas I have updated the PR based on your feedback.

am11 commented on src/tests/JIT/interpreter/InterpreterTester.cs:
```diff
@@ -16,15 +16,11 @@ public class InterpreterTester
-        startInfo.EnvironmentVariables["DOTNET_AltJitName"] = libInterp;
-        startInfo.EnvironmentVariables["DOTNET_AltJitPath"] = Path.Combine(coreRoot, libInterp);
-        startInfo.EnvironmentVariables["DOTNET_AltJit"] = "RunInterpreterTests";
+        var startInfo = new ProcessStartInfo(Path.Combine(coreRoot, "corerun"), interpreterApp);
```
Is .exe not needed on windows?
```suggestion
        string corerun = TestLibrary.Utilities.IsWindows ? "corerun.exe" : "corerun";
        var startInfo = new ProcessStartInfo(Path.Combine(coreRoot, corerun), interpreterApp);
```

janvorli replied:
No, it is not needed. It works both with and without it.

am11 replied:
It will at least keep things consistent with the existing practice; which may seem over-defensive or it's there to avoid unnecessary probes (CreateProcess will look for exact match, then fallback to appending .exe, .cmd .bat etc.)

```
src/libraries/System.Formats.Tar/tests/TarEntry/TarEntry.ExtractToFile.Tests.cs:            string linkTarget = PlatformDetection.IsWindows ? @"C:\Windows\system32\notepad.exe" : "/usr/bin/nano";
src/libraries/System.Formats.Tar/tests/TarEntry/TarEntry.ExtractToFileAsync.Tests.cs:                string linkTarget = PlatformDetection.IsWindows ? @"C:\Windows\system32\notepad.exe" : "/usr/bin/nano";
src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectory.Stream.Tests.cs:                entry.LinkName = PlatformDetection.IsWindows ? @"C:\Windows\System32\notepad.exe" : "/usr/bin/nano";
src/libraries/System.Formats.Tar/tests/TarFile/TarFile.ExtractToDirectoryAsync.Stream.Tests.cs:                    entry.LinkName = PlatformDetection.IsWindows ? @"C:\Windows\System32\notepad.exe" : "/usr/bin/nano";
src/tests/readytorun/coreroot_determinism/Program.cs:    public static string OSExeSuffix(string path) => (OperatingSystem.IsWindows() ? path + ".exe" : path);
src/tools/illink/test/ILLink.Tasks.Tests/ILLink.Tasks.Tests.cs:                 string dotnetToolName = OperatingSystem.IsWindows () ? "dotnet.exe" : "dotnet";
```

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -3749,7 +3928,7 @@ BOOL EEJitManager::GetBoundariesAndVars(
     } CONTRACTL_END;
 
-    CodeHeader * pHdr = GetCodeHeaderFromDebugInfoRequest(request);
+    TCodeHeader * pHdr = GetCodeHeaderFromDebugInfoRequest<TCodeHeader>(request);
```
`GetBoundariesAndVarsWorker` does not have to be template if we can change the argument for `GetBoundariesAndVarsWorker` to be `PTR_BYTE pDebugInfo`, move the `GetCodeHeaderFromDebugInfoRequest` to the caller.

janvorli replied:
Nice, I did the same to the GetRichDebugInfoWorker

janvorli replied:
Done

janvorli committed:
PR Feedback

* Get rid of some template usages by small refactoring

kg commented on src/coreclr/vm/eetwain.cpp:
```diff
@@ -2394,3 +2277,60 @@ ULONG32 EECodeManager::GetStackParameterSize(EECodeInfo * pCodeInfo)
+                                              unsigned        flags,
+                                              CodeManState   *pState)
+{
+    // Interpreter-TODO: Implement this
```
Is there a specific reason why these aren't `ASSERTE(false)` like the other interpreter TODOs?

janvorli replied:
I didn't add assert to the ones that I knew we need to implement and that I am already implementing in a follow up change to enable stack walking with interpreted frames on stack. For the ones with the assert, those are the ones that I have thought that won't likely be implemented for the interpreter.

kg commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -2428,7 +2444,15 @@ HeapList* LoaderCodeHeap::CreateCodeHeap(CodeHeapRequestInfo *pInfo, LoaderHeap
+    if (pInfo->IsInterpreted())
+    {
+        pHp->CLRPersonalityRoutine = NULL;
+        pCodeHeap->m_LoaderHeap.ReservePages(1);
```
Is this important? I assume the idea is to ensure that the heap has something in it?

janvorli replied:
Yes, the code further requires at least one reserved page to be present. For the personality routine case in the else branch, the `AllocMem` ensures that (we alloc hence a page gets reserved due to that).  
There is a comment in the #else below explaining it for the case of targets that don't use personality routines.

kg commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -3532,7 +3661,52 @@ void EEJitManager::RemoveCodeHeapFromDomainList(CodeHeap *pHeap, LoaderAllocator
+{
+    STANDARD_VM_CONTRACT;
+
+    // If the JIT is already loaded, don't take the lock.
```
nitpick: we say 'jit' and 'interpreter' interchangeably a few times in this function, i assume because it's based on altjit loader logic

janvorli replied:
That's a copy & paste issue, I'll fix it. This code is a clone of the `EEJitManager::LoadJIT`

kg commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -3532,7 +3661,52 @@ void EEJitManager::RemoveCodeHeapFromDomainList(CodeHeap *pHeap, LoaderAllocator
+    ICorJitCompiler* newInterpreter = NULL;
+    m_interpreter = NULL;
+
+    g_interpreterLoadData.jld_id = JIT_LOAD_MAIN;
```
Why is it `JIT_LOAD_MAIN` instead of a new `JIT_LOAD_INTERP`? I see `JIT_LOAD_ALTJIT` exists. (I don't know the significance of jld_id, I can't find any code that actually reads it, so maybe it doesn't matter?)

janvorli replied:
I think this is just a debugging mean, I'll need to ask JIT folks to get a definitive answer. Anyways, I will add a new JIT_LOAD_variant.

kg commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -4624,12 +4980,20 @@ BOOL ExecutionManager::IsManagedCodeWorker(PCODE currentPC, RangeSectionLockStat
     if (pRS == NULL)
```
In the context of this function, currentPC isn't necessarily the eip register or equivalent, but instead a code address that may or may not be managed code, right? Or will we eventually need to handle 'currentPC is inside of the interpreter's native implementation, so we need to find the interpreter stack and fish out its IP'? I can't tell from grepping around whether that is potentially going to happen here.

(If it does matter for later, it'd just be cool to see a comment pointing out that problem or an interpreter-todo like the other ones you've added.)

janvorli replied:
For the interpreter, PC always represents the interpreter PC, that means a byte code address pointer. That's the way everywhere. I see the interpreter byte code as an instruction set of a virtual processor.

kg commented on src/coreclr/vm/codeman.h:
```diff
@@ -247,12 +259,92 @@ typedef struct _hpCodeHdr
+    VOID* GetCalledMethods()
+    {
+        SUPPORTS_DAC;
+        return pRealCodeHeader->pCalledMethods;
```
pCalledMethods isn't defined in interpreter real code header is it? copy-paste error?

janvorli replied:
I used to have that on the code header even for the interpreter, but removed it later based on a feedback. I'll delete this leftover.

kg commented on src/coreclr/vm/codeman.h:
```diff
@@ -1711,9 +1942,18 @@ class EEJitManager : public IJitManager
+    VPTR_VTABLE_CLASS(EEJitManager, EECodeGenManager)
 
 public:
+    virtual ICorJitCompiler* GetCompiler()
```
Is there a technical reason why we're not using `override` in cases like this where we're implementing an abstract virtual from the parent inside of a final class? Or is it just a style thing?

janvorli replied:
Yes, adding it to even a single method triggers the compiler to require it everywhere in the class and if I remember it correctly, even on the base class. And I think it has some other ripple effect too which I cannot recall now.

janvorli commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -3532,7 +3661,52 @@ void EEJitManager::RemoveCodeHeapFromDomainList(CodeHeap *pHeap, LoaderAllocator
+    LoadAndInitializeJIT(ExecutionManager::GetInterpreterName() DEBUGARG(interpreterPath), &m_interpreterHandle, &newInterpreter, &g_interpreterLoadData, getClrVmOs());
+
+    // Publish the interpreter.
+    m_interpreter = newInterpreter;
```
The critical section is sufficient. All threads go through this InterpreterJitManager::LoadInterpreter before they ever try to touch it.

jkotas replied:
> The critical section is sufficient. All threads go through this InterpreterJitManager::LoadInterpreter before they ever try to touch it.

LoadInterpreter checks whether the interpreter is loaded outside the critical section using IsInterpreterLoaded. It means that this check can pass, but the interpreter may not be actually loaded on the current thread due hardware or compiler memory access reordering.

As @kg pointed out, this can use some memory barriers to be 100% correct when multithreading is involved.

kg commented on src/coreclr/vm/codeman.h:
```diff
@@ -2449,6 +2627,141 @@ class ReadyToRunJitManager final: public IJitManager
+    virtual DWORD GetCodeType()
+    {
+        LIMITED_METHOD_DAC_CONTRACT;
+        return (miManaged | miIL);
```
Is this what we want? Grepping around it seems like some callers will assume that `(miManaged | miIL)` means JITcode, and the DAC will also treat this as JITcode I think. Vs if we returned something else the DAC would see it as UNKNOWN (what we want I think?)

i.e. `PublishUnwindInfoForExistingMethods` looks for this, I'm not sure whether interpreter IR would strictly need unwind info in the same way or if it would just be polluting a table that should only contain native executable code (JIT or R2R, regardless) that will be on the stack during an unwind?

janvorli replied:
Interpreter won't use any unwind info and the PublishUnwindInfoForExistingMethods is not called for the interpreter code. 
It seems that the miManaged and miIL needs to be there as it is both correct. I was thinking about adding some extra flag for interpreter code, but it is problematic as the code type has only two bits reserved and the CorMethodImpl is something that ILASM sets too. So I assume C# compiler would also set it. I didn't want to require tools to need to  be modified.
So far, I haven't hit any issue with the current state, but it is theoretically possible that we will need to revisit this eventually. I'll add a TODO here.

jkotas replied:
> it is problematic as the code type has only two bits reserved and the CorMethodImpl is something that ILASM sets too.

This method is only used in DAC and in few asserts. It is an anti-parent that it reuses CorMethodImpl enum that is part of the ECMA metadata definition. The value returned by this method has nothing to do with the ECMA metadata. We can change it to use its own enum with as many bits as we need if necessary.

kg commented on src/coreclr/vm/codeman.h:
```diff
@@ -2449,6 +2627,141 @@ class ReadyToRunJitManager final: public IJitManager
+
+    virtual void JitTokenToMethodRegionInfo(const METHODTOKEN& MethodToken, MethodRegionInfo * methodRegionInfo)
+    {
+        // Not used for the interpreter
```
maybe make this an interpreter-todo as well like the others

kg commented on src/coreclr/vm/codeman.h:
```diff
@@ -2594,6 +2910,37 @@ ULONG       GetFixedStackSize();
+{
+    LIMITED_METHOD_DAC_CONTRACT;
+    _ASSERTE(methodStartAddress != (TADDR)NULL);
+    ARM_ONLY(_ASSERTE((methodStartAddress & THUMB_CODE) == 0));
```
nitpick: This looks copy-pasted from the JIT version and could bite us if we ever make interpreter opcodes single-byte.
Alternately, we could explicitly enforce the alignment of interpreter IR here (i think it's 4 bytes right now?)

janvorli replied:
Right, this assert should not be here.

kg commented on src/coreclr/vm/jitinterface.h:
```diff
@@ -565,20 +565,220 @@ class CEEInfo : public ICorJitInfo
+                        );
+
+    EECodeGenManager*       m_jitManager;   // responsible for allocating memory
+    void*                   m_CodeHeader;   // descriptor for JITTED code - read/execute address
```
This is no longer restricted to jitted code, right? It also will be there for interpreted code, just not R2R?

janvorli replied:
Right, this is used for the interpreted code too.

kg commented on src/coreclr/vm/jitinterface.h:
```diff
@@ -565,20 +565,220 @@ class CEEInfo : public ICorJitInfo
+
+    EECodeGenManager*       m_jitManager;   // responsible for allocating memory
+    void*                   m_CodeHeader;   // descriptor for JITTED code - read/execute address
+    void*                   m_CodeHeaderRW; // descriptor for JITTED code - code write scratch buffer address
```
vs this one which if I recall correctly *is* only for jitcode, because it's a temporary allocation we make if the code header is in RX memory instead of RWX or RW?

janvorli replied:
For the interpreter, the m_CodeHeader (that's the RX one for JIT) and m_CodeHeaderRW have the same value. It simplifies the change, since the code that's common for the interpreter and jit needs to access through a writeable address.

kg commented on src/coreclr/vm/jitinterface.h:
```diff
@@ -713,7 +873,7 @@ class CEEJitInfo : public CEEInfo
     }
 
-    BOOL JitAgain()
+    virtual BOOL JitAgain() override
```
nitpick: i think `virtual override` is redundant and we only need `override`?

janvorli replied:
Right, will fix

kg commented on src/coreclr/vm/jitinterface.h:
```diff
@@ -931,50 +1046,57 @@ protected :
+            GC_NOTRIGGER;
+        } CONTRACTL_END;
+
+        CEECodeGenInfo::ResetForJitRetry();
```
Do we want to do ASSERTE(false) here and/or an interpreter todo instead? I don't fully understand what jit retry implies or whether there is some scenario where the interpreter would want to do it, but it feels like something the interpreter wouldn't do.

It looks like jit retry is an altjit only thing, to begin with?

janvorli replied:
This is a cleanup when jitting (even interpreter jitting) fails. It clears the state of the CEECodeGenInfo so that it can be reused.

kg commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -11085,19 +11101,190 @@ PatchpointInfo* CEEJitInfo::getOSRInfo(unsigned* ilOffset)
+    {
+        COMPlusThrowHR(CORJIT_OUTOFMEM);
+    }
+    if (ETW_EVENT_ENABLED(MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context, MethodJitMemoryAllocatedForCode))
```
Do we want to reuse the jit ETW events for interp? Or define new ones, or leave them out? Might be good to have an interpreter todo here if you think we'll want to update this later

janvorli committed:
Reflect feedback

janvorli committed:
Merge branch 'main' into interpreter-code-manager

janvorli commented:
/ba-g the wasm failure is unrelated and it is a known infra issue.
